### PR TITLE
Readonly pipe structs and remove lenght

### PR DESCRIPTION
**Pull request by pakrym**
Does not affect the throughput.

### CODE REVIEW

**Comment by YohDeadfall** [comment]
Would be `ReadCursor` replaced by `Position` from sequences?

**Comment by pakrym** [comment]
@YohDeadfall it's still being discussed.

File: src/System.IO.Pipelines/System/IO/Pipelines/PreservedBuffer.cs [hunk 1]
```diff
@@ -39,8 +39,6 @@ public void Dispose()
                     break;
                 }
             }
-
-            _buffer.ClearCursors();
```
Comment (davidfowl) [inline src/System.IO.Pipelines/System/IO/Pipelines/PreservedBuffer.cs]: Unrelated?

File: src/System.IO.Pipelines/System/IO/Pipelines/ReadableBuffer.cs [hunk 2]
```diff
@@ -9,24 +9,23 @@ namespace System.IO.Pipelines
...
         /// Length of the <see cref="ReadableBuffer"/> in bytes.
         /// </summary>
-        public long Length => BufferLength;
+        public long Length => BufferStart.GetLength(BufferEnd);
 
         /// <summary>
         /// Determines if the <see cref="ReadableBuffer"/> is empty.
         /// </summary>
-        public bool IsEmpty => BufferLength == 0;
+        public bool IsEmpty => Length == 0;
```
Comment (benaadams) [inline src/System.IO.Pipelines/System/IO/Pipelines/ReadableBuffer.cs]: Does removing actual length cause issue for `IsEmpty` as its now a calculation? 

Can it just check the first segment rather than full chain?

Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/ReadableBuffer.cs]: We need to figure out some issues with cursors and switch to cursor comparison for `IsEmpty`
